### PR TITLE
fix: [ASR->LLVM] use pointer to cptr when unspecified intent

### LINKED ISSUE #4215: argument type mismatch when absent Value attribute for a C-interfaced subroutine
### MRE
For the below fortran program `main.f90`:
```f90
program main
    use iso_c_binding, only : c_ptr, c_f_pointer, c_int
    implicit none

    interface
        subroutine allocate_array(ptr, size) bind(C)
            use iso_c_binding, only : c_ptr, c_int
            type(c_ptr) :: ptr
            integer(c_int), value :: size
        end subroutine allocate_array
    end interface

    type(c_ptr) :: c_pointer
    integer(c_int) :: size

    size = 10
    call allocate_array(c_pointer, size)
end program main
```

with C function defined in `allocate.c` file as:
```c
#include <stdlib.h>

void allocate_array(int **ptr, int size) {
    *ptr = (int *)malloc(size * sizeof(int));
    for (int i = 0; i < size; i++) {
        (*ptr)[i] = i;
    }
}

```

On compilation with LFortran, using the below commands:
```
> gcc -c allocate.c -o allocate.o
> lfortran main.f90 allocate.o -L -o main.out
```

 I get an error:

```llvm
; ModuleID = 'LFortran'
source_filename = "LFortran"

declare void @allocate_array(void*, i32)

define i32 @main(i32 %0, i8** %1) {
.entry:
  call void @_lpython_call_initial_functions(i32 %0, i8** %1)
  %c_pointer = alloca void*, align 8
  %size = alloca i32, align 4
  store i32 10, i32* %size, align 4
  %2 = load i32, i32* %size, align 4
  call void @allocate_array(void** %c_pointer, i32 %2)
  ret i32 0
}

declare void @_lpython_call_initial_functions(i32, i8**)
asr_to_llvm: module failed verification. Error:
Call parameter type does not match function signature!
  %c_pointer = alloca void*, align 8
 void*  call void @allocate_array(void** %c_pointer, i32 %2)

code generation error: asr_to_llvm: module failed verification. Error:
Call parameter type does not match function signature!
  %c_pointer = alloca void*, align 8
 void*  call void @allocate_array(void** %c_pointer, i32 %2)



Note: Please report unclear, confusing or incorrect messages as bugs at
https://github.com/lfortran/lfortran/issues.
```

While compilation with GFortran succeeds without any error (or output in this case)

### NOTE

interestingly, if I append `value` attribute (see in the code below for where it's added) to the declaration of `c_ptr` i.e. with the below fortran program:
```f90
program main
    use iso_c_binding, only : c_ptr, c_f_pointer, c_int
    implicit none

    interface
        subroutine allocate_array(ptr, size) bind(C)
            use iso_c_binding, only : c_ptr, c_int
            type(c_ptr), value :: ptr  ! appended 'value' attribute
            integer(c_int), value :: size
        end subroutine allocate_array
    end interface

    type(c_ptr) :: c_pointer
    integer(c_int) :: size

    size = 10
    call allocate_array(c_pointer, size)
end program main
```

The compilation to show LLVM IR succeeds:
```llvm
; ModuleID = 'LFortran'
source_filename = "LFortran"

declare void @allocate_array(void*, i32)

define i32 @main(i32 %0, i8** %1) {
.entry:
  call void @_lpython_call_initial_functions(i32 %0, i8** %1)
  %c_pointer = alloca void*, align 8
  %size = alloca i32, align 4
  store i32 10, i32* %size, align 4
  %2 = load void*, void** %c_pointer, align 8
  %3 = load i32, i32* %size, align 4
  call void @allocate_array(void* %2, i32 %3)
  ret i32 0
}
```

### PR DESCRIPTION
**Pull request by gxyd**
fixes #4215

### CODE REVIEW

File: integration_tests/bindc_01b.f90 [hunk 1]
```diff
@@ -8,6 +8,11 @@ subroutine ret_ptr(p) bind(c, name="ret_ptr_c")
     type(c_ptr), intent(out) :: p
     end subroutine
 
+    subroutine ret_ptr1(p) bind(c, name="ret_ptr_c")
+    import :: c_ptr
+    type(c_ptr) :: p
+    end subroutine
```
Comment (Pranavchiku) [inline integration_tests/bindc_01b.f90:14]: What does this do?

Comment (gxyd) [reply integration_tests/bindc_01b.f90:14]: Same as `ret_ptr` already defined above, only difference being `p`'s intent is unspecified in this subroutine. Maybe I should use a difference name actually. `ret_ptr_unspecified_intent`

Comment (gxyd) [reply integration_tests/bindc_01b.f90:14]: Or atleast I'll put a comment describing the difference

**Review by certik — APPROVED** [review]
I think that's fine. @Pranavchiku can you also please review?

**Review by Pranavchiku — APPROVED** [review]

**Comment by Pranavchiku** [comment]
can you please rebase it, we can then merge (squash) it, if that is fine?

**Comment by gxyd** [comment]
I've directly squashed and merged it (wasn't necessary to rebase, was it?)

**Comment by Pranavchiku** [comment]
That's fine, if there are any significant changes in main that can break things in this PR we should prefer rebasing it on top to assure that CI at main never fails.

**Pull request merged**